Hw3 branch

CoreyPatrickPDX opened the pull request:
Completed homework 3. Edited files for:
Part 1:
    -01-characters.html: Added some bootstrap styling for the main and section tags.
    -01-characters.js: implemented code to display all characters and their images (with alt-text)
Part 2: 
    -02-charts.html: added styling for the div container holding the chart
    -02-charts.js: implemented code to parse the api data for character families and how many of each. Checked for typos in family names and blank to values I decided.
Part 3: Created a react-app with create-react-app which added multiple folders and files some of which I edited, then I added multiple on top including:
    -src/navigation/navbar.js: implements a navbar that will be shared in each page
    -src/utils/chartColors.js: The color arrays used in part 2 for the charts that will be imported into the houses page
    -views/home.js: a simple home page using react
    -views/houses.js: this is basically part 2 of the homework but using react to create the page instead
    -views/search.js: a page that allows a user to search for a specific character using their full name. Displays their name and picture on return.
    -App.js: Edited to show the navbar on each rendered page

CoreyPatrickPDX committed:
finished 01 and 02 of hw3, started the react portion

CoreyPatrickPDX committed:
finished work on part 3 for the react site

CoreyPatrickPDX committed:
Forgot to take out a console.log statement

CoreyPatrickPDX committed:
Some small cleanup, word changes.

CoreyPatrickPDX committed:
Updated search page to have alt text with the character image that holds thier full name

CoreyPatrickPDX committed:
added more colors to 02-charts.js for the chart

stevengborrego reviewed (commented):
Hi Corey, everything looks really great! I'm seeing a couple odd stylings though in the react app. The chart is quite large, and the search feature displays to the left instead of the center. Not sure if these were intentional stylings, but I think it might benefit from downsizing the chart and centering the search result. Either way, everything looks really good. Great job!

CoreyPatrickPDX committed:
Made the change to the chart size as Nicholas suggested. Also, centered the image and text from the search function so it looks a little better. Changed the search function to accept lower case values.

CoreyPatrickPDX commented:
Made some style changes that Steven suggested, not Nicholas (my bad). Updated the view on the search page to center the character picture and full name, also styled the chart to be a bit smaller so the whole thing fits on the page. I also updated the search feature to accept lower case values from the user instead of only capitalized.

JoshuaWachana commented on hw3/01-characters.js:
```diff
@@ -1,4 +1,44 @@
+    const row = document.createElement('div');
+    row.classList.add('row', 'g-2');
+
+    data.forEach((character) => {
+      const element = document.createElement('div');
+      element.classList.add('col-sm-6', 'col-md-3');
+
+      const charDiv = document.createElement('div');
+      charDiv.classList.add('text-center', 'p-2', 'character');
+
+      const image = document.createElement('img');
+      image.src = character.imageUrl;
+      image.alt = character.fullName;
+      image.classList.add('charImage', 'img-fluid');
+
+      const charName = document.createElement('h4');
+      charName.textContent = character.fullName;
+      charName.classList.add('w-75', 'mx-auto');
+
+      const charTitle = document.createElement('h5');
+      charTitle.innerHTML = character.title;
+      charTitle.classList.add('w-75', 'mx-auto');
+
+      charDiv.append(image);
+      charDiv.append(charName);
+      charDiv.append(charTitle);
+      element.append(charDiv);
+      row.append(element);
+    });
+    app.append(row);
+  })
```
This is a lot of code to put inside a forEach().

Consider creating an additional function which would be responsible for creating and appending elements "characters" to the DOM. That new function would have the dedicated task of taking the data and processing it into into elements to add to the DOM. The function name would match it's responsibility too, for example addCharactersToDom(). So in the future when someone wants to change how the data is displayed on the page it would be easy for them to know to focus on that function instead of looking through the forEach and modifying things in there. It's more about separation of concerns and making it readable.

JoshuaWachana commented on hw3/01-characters.js:
```diff
@@ -1,4 +1,44 @@
+    row.classList.add('row', 'g-2');
+
+    data.forEach((character) => {
+      const element = document.createElement('div');
+      element.classList.add('col-sm-6', 'col-md-3');
+
+      const charDiv = document.createElement('div');
+      charDiv.classList.add('text-center', 'p-2', 'character');
+
+      const image = document.createElement('img');
+      image.src = character.imageUrl;
+      image.alt = character.fullName;
+      image.classList.add('charImage', 'img-fluid');
+
+      const charName = document.createElement('h4');
+      charName.textContent = character.fullName;
+      charName.classList.add('w-75', 'mx-auto');
+
+      const charTitle = document.createElement('h5');
+      charTitle.innerHTML = character.title;
+      charTitle.classList.add('w-75', 'mx-auto');
```
Consider using the figure and figcaption elements, this exercise would be a good use case for them.

JoshuaWachana replied:
![Screenshot (268)](https://github.com/CoreyPatrickPDX/frontend-homework/assets/38127011/909ff935-38fc-4d8e-90a6-e72a07a70a58)



There are some accessibility issues when the character is hovered over. There are contrast errors.

JoshuaWachana commented on hw3/hw3-react/src/navigation/navbar.js:
```diff
@@ -0,0 +1,60 @@
+          <span class="navbar-toggler-icon"></span>
+        </button>
+        <div class="collapse navbar-collapse" id="navbarNav">
+          <ul class="navbar-nav">
+            <li class="nav-item active">
+              <a class="nav-link" href="/">
+                Home
+              </a>
+            </li>
+            <li class="nav-item">
+              <a class="nav-link" href="/search">
+                Search
+              </a>
+            </li>
+            <li class="nav-item">
+              <a class="nav-link" href="/houses">
+                Houses
+              </a>
+            </li>
+          </ul>
+        </div>
+      </nav>
+      <Routes>
+        <Route path="/" element={<Home title=" Game of Thrones Enthusiast" />} />
+
+        <Route
+          path="/search"
+          element={<Search title="Search for your favorite character" />}
+        />
+
+        <Route
+          path="/houses"
+          element={<Houses title="Game of Thrones Family/House Distribution" />}
+        />
+      </Routes>
+    </div>
+  );
+}
+
+export default Navbar;
```
Why are the Home, Search and Houses components children of the Navbar component? The Navbar component should be used to build the navbar, not the majority of the site.

JoshuaWachana commented on hw3/hw3-react/src/views/houses.js:
```diff
@@ -0,0 +1,76 @@
+  margin: '20px auto',
+};
+
+function Houses(props) {
+  const [posts, setPosts] = useState([]);
+  const { title } = props;
+  const url = 'https://thronesapi.com/api/v2/Characters';
+  const houseNames = [];
+  const houseValues = [];
+
+  useEffect(() => {
+    const fetchPost = async () => {
+      let response = await axios.get(url);
+      setPosts(response.data);
+    };
+    fetchPost();
+  });
```
Calls the same API and fetches the same data as that of Search.js

JoshuaWachana commented on hw3/hw3-react/src/views/search.js:
```diff
@@ -0,0 +1,79 @@
+  let [pic, setPic] = useState('');
+  let [alt, setAlt] = useState('');
+
+  const url = 'https://thronesapi.com/api/v2/Characters';
+  useEffect(() => {
+    fetch(url)
+      .then((res) => res.json())
+      .then((characters) => {
+        setData(characters);
+        setIsLoaded(true);
+      })
+      .catch((error) => console.log(error));
+  }, []);
```
Calls the same API and fetches the same data as that of Houses.js

Fetching the data (calling the exact same API) twice, in both the Houses and Search components is inefficient. The data they both rely on is the same, and additionally both of these components have the same parent (should be something like App.js, not Navbar).

So why not let the data flow from parent to child?

It would make a lot more sense to call the API once in the App component, then pass the data as props to the houses and search components. This would be more efficient since the API would only need to be called once, but the data would be available across the components that need it.

In summary, you can just fetch the data once and then pass it to child components via props.
